### PR TITLE
ARROW-2512: [WIP] [Python] Using Pytorch as Python interface for GPU objects

### PR DESCRIPTION
**Pull request by Wapaul1**
This is currently a work in progress. Pytorch 0.4.0 is required for this to work.

## Instructions
First, go to `arrow/python/pytorch_example` and run `python setup.py install`, which will install the extension to pytorch. Then build arrow normally with plasma and gpu enabled as below:
```
cmake -DCMAKE_BUILD_TYPE=$ARROW_BUILD_TYPE \
      -DCMAKE_INSTALL_PREFIX=$ARROW_HOME \
      -DARROW_PYTHON=on \
      -DARROW_PLASMA=on \
      -DARROW_BUILD_TESTS=OFF \
      -DARROW_GPU=ON \
      ..
```
and build pyarrow with `python setup.py build_ext --build-type=$ARROW_BUILD_TYPE        --with-plasma --inplace` (this may not be necessary, this is just how I installed pyarrow).

Then in ipython, create a plasma store and a client:
```python
import os
import random
import shutil
import signal                       
import subprocess
import tempfile
import time
 
import numpy as np
import pyarrow as pa
import pandas as pd
 
 
DEFAULT_PLASMA_STORE_MEMORY = 10 ** 8
USE_VALGRIND = os.getenv("PLASMA_VALGRIND") == "1"
 
 
def random_name():
    return str(random.randint(0, 99999999))
def start_plasma_store(plasma_store_memory=DEFAULT_PLASMA_STORE_MEMORY,
                        stdout_file=None, stderr_file=None):
     plasma_store_executable = os.path.join(pa.__path__[0], "plasma_store")
     plasma_store_name = "/tmp/plasma_store{}".format(random_name())
     command = [plasma_store_executable,
               "-s", plasma_store_name,
               "-m", str(plasma_store_memory)]
     pid = subprocess.Popen(command, stdout=stdout_file, stderr=stderr_file)
     time.sleep(0.1)
     return plasma_store_name, pid
import pyarrow.plasma as plasma
        # Start Plasma store.
plasma_store_name, p = start_plasma_store()
client = plasma.connect(plasma_store_name, "", 64)

```

Now simply create a numpy float32 array, and pass it to `client.gpu_example`
```
arr = np.random.rand(20,20).astype('float32')
result = client.gpu_example(arr)
```
which should return a pytorch tensor.

### CODE REVIEW

**Comment by pitrou** [comment]
From a cursory look, I think there are some flaws here:
1) GPU support in Python shouldn't require PyTorch (last I tried, I think I didn't even manage to install a working PyTorch package -- regardless, it shouldn't be a dependency for this)
2) in Python, PlasmaBuffer is mostly a relic: it will go when [ARROW-2363](https://issues.apache.org/jira/browse/ARROW-2363) is solved

Also, IMHO [ARROW-2447](https://issues.apache.org/jira/browse/ARROW-2447) should get implemented before this moves forward, to avoid accumulating technical debt.

**Comment by pcmoritz** [comment]
@pitrou Thanks for the feedback! You are right, we need to do some refactoring before this gets merged. Also pytorch won't be a dependency of plasma, the plan is to make it to have an example that shows how to load a tensor from plasma GPU memory directly into pytorch.

File: cpp/src/plasma/format/plasma.fbs [hunk 1]
```diff
@@ -215,8 +215,9 @@ table PlasmaStatusRequest {
 }
 
 enum ObjectStatus:int {
+  Null = 0,
```
Comment (pcmoritz) [inline cpp/src/plasma/format/plasma.fbs]: This is now fixed in the latest master, if you rebase or merge the master into this branch it should work!

**Comment by wesm** [comment]
This PR is more than a year old, I'm closing but feel free to submit a new PR after addressing feedback and rebase issues

**Pull request closed without merging**